src/internal/client.ts

```diff
@@ -740,4 +743,63 @@ export class TypedClient {
       (err) => cb(err),
     )
   }
+
+  /**
+   * Get part-info of all parts of an incomplete upload specified by uploadId.
+   */
+  protected async listParts(bucketName: string, objectName: string, uploadId: string): Promise<UploadedPart[]> {
+    if (!isValidBucketName(bucketName)) {
+      throw new errors.InvalidBucketNameError('Invalid bucket name: ' + bucketName)
+    }
+    if (!isValidObjectName(objectName)) {
+      throw new errors.InvalidObjectNameError(`Invalid object name: ${objectName}`)
+    }
+    if (!isString(uploadId)) {
+      throw new TypeError('uploadId should be of type "string"')
+    }
+    if (!uploadId) {
+      throw new errors.InvalidArgumentError('uploadId cannot be empty')
+    }
+
+    const parts: UploadedPart[] = []
+    let marker = 0
+    let result
+    do {
+      result = await this.listPartsQuery(bucketName, objectName, uploadId, marker)
+      marker = result.marker
+      parts.push(...result.parts)
+    } while (result.isTruncated)
+
+    return parts
+  }
+
+  /**
+   * Called by listParts to fetch a batch of part-info
+   */
+  private async listPartsQuery(bucketName: string, objectName: string, uploadId: string, marker: number) {
+    if (!isValidBucketName(bucketName)) {
+      throw new errors.InvalidBucketNameError('Invalid bucket name: ' + bucketName)
+    }
+    if (!isValidObjectName(objectName)) {
+      throw new errors.InvalidObjectNameError(`Invalid object name: ${objectName}`)
+    }
+    if (!isString(uploadId)) {
+      throw new TypeError('uploadId should be of type "string"')
+    }
+    if (!isNumber(marker)) {
+      throw new TypeError('marker should be of type "number"')
+    }
+    if (!uploadId) {
+      throw new errors.InvalidArgumentError('uploadId cannot be empty')
+    }
+
+    let query = `uploadId=${uriEscape(uploadId)}`
+    if (marker) {
+      query += `&part-number-marker=${marker}`
+    }
```

Comment:
We should probably check if `marker` value is negative, because I don't think it can be a negative value.

Reply:
I'm not quite sure about this, this mark should be valid if it's not the first request.

Reply:
I will check this.

Reply:
the `marker` here is read from `marker = result.marker`. and loop will stop if the reponse is not  truncated. So only valid marker (except initial value `0`) will be passed to this method.

Reply:
S3 documents say the marker is valid if `isTruncated` is true.

Reply:
patch applied

Reply:
Thank you for incorporating